### DISCUSSION TITLE
Some new guidelines that will help lessen merge conflicts

### DISCUSSION THREAD

**Discussion by brittanyarima** [brittanyarima/Steps#69] [Announcements]
1.Before adding a file, commit your changes, 
2. then pull from the master so that you have the latest.If someone has added a file, you now have the latest .pbxproj
3. Add your file(s).
4. Immediately commit and push your changes up to the master (hopefully, before another collaborator has added another file).

**Comment by Drag0ndust** [comment]
I think you wanted to write the "dev" branch not the "master" in your message above.
To get rid of the problems with the Xcode project files, you could use a tool which generates the Xcode files for you, e.g. `Tuist` or `xcodegen`

**Reply by brittanyarima** [reply]
Oh thanks! I'll look into Tuist this week!

**Reply by Drag0ndust** [reply]
If you need help ping me, I can also have a look

**Comment by DandyLyons** [comment]
I'd be interested in learning more about `Tuist` and `xcodegen` like @Drag0ndust says. I've been hearing about xcode developers complain about merge conflicts of xcode projects for years and now I'm experiencing it for the first time. And I've heard of them using `Tuist` and `xcodegen` precisely to solve that problem. 

Another solution that I've heard gain a lot of steam is Swift Packages. One approach is to put practically everything in a Swift Package. That way there's no messy complexity to a pbxproj file. (There are some caveats though. For example, Swift Packages cannot build into an actual app, so you still need an Xcode project to hold the swift package(s).)
Here's a talk I found yesterday that talks about the SPM approach: https://youtu.be/9KOnO77msU8?feature=shared&t=116

@Drag0ndust do you have experience/resources regarding setting up `Tuist` or `xcodegen`?

**Reply by Drag0ndust** [reply]
@DandyLyons I use Tuist in my own projects and also in some customer projects. Sometimes it is a little bit tricky, but it is definitely worth to use it.

**Reply by DandyLyons** [reply]
Do you have a tutorial that you'd recommend?

**Reply by Drag0ndust** [reply]
not really. Just work through the [Getting started](https://docs.tuist.io/tutorial/get-started/)

**Comment by DandyLyons** [comment]
@brittanyarima Here's a free video from Pointfree.co on modularizing code. I highly recommend it. But they show the pros/cons of different approaches and how using Swift Packages allows you to minimize changes to your xcode project which minimizes the creation of merge conflicts. 

https://www.pointfree.co/episodes/ep171-modularization-part-1#t1373
